// jovo-platform-facebookmessenger/dist/src/Enums.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var MessageType;
(function (MessageType) {
    MessageType["Response"] = "RESPONSE";
    MessageType["Update"] = "UPDATE";
    MessageType["MessageTag"] = "MESSAGE_TAG";
})(MessageType = exports.MessageType || (exports.MessageType = {}));
var AttachmentType;
(function (AttachmentType) {
    AttachmentType["Audio"] = "audio";
    AttachmentType["Video"] = "video";
    AttachmentType["Image"] = "image";
    AttachmentType["File"] = "file";
})(AttachmentType = exports.AttachmentType || (exports.AttachmentType = {}));
var TemplateType;
(function (TemplateType) {
    TemplateType["Generic"] = "generic";
    TemplateType["Button"] = "button";
    TemplateType["Receipt"] = "receipt";
    TemplateType["Airline"] = "airline_boardingpass";
    TemplateType["Media"] = "media";
})(TemplateType = exports.TemplateType || (exports.TemplateType = {}));
var ButtonType;
(function (ButtonType) {
    ButtonType["Link"] = "web_url";
    ButtonType["Postback"] = "postback";
    ButtonType["Call"] = "phone_number";
    ButtonType["Login"] = "account_link";
    ButtonType["Logout"] = "account_unlink";
    ButtonType["Game"] = "game_play";
})(ButtonType = exports.ButtonType || (exports.ButtonType = {}));
var WebViewHeightRatio;
(function (WebViewHeightRatio) {
    WebViewHeightRatio["Compact"] = "COMPACT";
    WebViewHeightRatio["Tall"] = "TALL";
    WebViewHeightRatio["Full"] = "FULL";
})(WebViewHeightRatio = exports.WebViewHeightRatio || (exports.WebViewHeightRatio = {}));
var MediaType;
(function (MediaType) {
    MediaType["Image"] = "image";
    MediaType["Video"] = "video";
})(MediaType = exports.MediaType || (exports.MediaType = {}));
var QuickReplyContentType;
(function (QuickReplyContentType) {
    QuickReplyContentType["Text"] = "text";
    QuickReplyContentType["PhoneNumber"] = "user_phone_number";
    QuickReplyContentType["Email"] = "user_email";
})(QuickReplyContentType = exports.QuickReplyContentType || (exports.QuickReplyContentType = {}));
var SenderActionType;
(function (SenderActionType) {
    SenderActionType["MarkSeen"] = "mark_seen";
    SenderActionType["TypingOn"] = "typing_on";
    SenderActionType["TypingOff"] = "typing_off";
})(SenderActionType = exports.SenderActionType || (exports.SenderActionType = {}));
